sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -740,6 +813,9 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
   /** Binds an input expression to a given input schema */
   protected def bind(in: InType, inputSchema: Seq[Attribute]): InType
 
+  /** Generate the code to be compiled represented as string */
+  def generateCodeAndComments(ctx: CodegenContext, in: InType): CodeAndComment
```

Comment:
I added this method for testability. We can test whether place holder is put in generated code as we intended.